src/java/org/apache/cassandra/service/accord/txn/TxnNamedRead.java

```diff
@@ -122,24 +149,68 @@ public AsyncChain<Data> read(boolean isForWriteTxn, SafeCommandStore safeStore,
         // immediately after the transaction executed, and this simplifies things a great deal
         int nowInSeconds = (int) TimeUnit.MICROSECONDS.toSeconds(executeAt.hlc());
 
+        if (ConsensusRequestRouter.instance.isKeyInMigratingRangeFromPaxos(tms, key))
+            return performCoordinatedRead(consistencyLevel, command, nowInSeconds, metrics);
+        else
+            return performLocalRead(command, nowInSeconds);
+    }
+
+    private AsyncChain<Data> performCoordinatedRead(ConsistencyLevel consistencyLevel, SinglePartitionReadCommand command, int nowInSeconds, AccordClientRequestMetrics metrics)
```

Comment:
I think it would be possible to read data from 'the future'. If we read from a replica that has already applied this operation, we will break correctness. This is especially possible in recovery.

Reply:
I think you are right. There is no path forward that doesn't end with the coordinated reads being done from inside Accord through command stores.